src/mlpack/methods/hmm/hmm_impl.hpp

```diff
@@ -643,23 +654,48 @@ void HMM<Distribution>::Backward(const arma::mat& dataSeq,
   }
 }
 
+template<typename Distribution>
+void HMM<Distribution>::ConvertToLogSpace() const {
+    logInitial = log(initialProxy);
+    logTransition = log(transitionProxy);
+}
+
 //! Serialize the HMM.
 template<typename Distribution>
 template<typename Archive>
-void HMM<Distribution>::serialize(Archive& ar, const unsigned int /* version */)
+void HMM<Distribution>::load(Archive& ar, const unsigned int /* version */)
 {
+  arma::mat transition;
+  arma::vec initial;
   ar & BOOST_SERIALIZATION_NVP(dimensionality);
   ar & BOOST_SERIALIZATION_NVP(tolerance);
   ar & BOOST_SERIALIZATION_NVP(transition);
   ar & BOOST_SERIALIZATION_NVP(initial);
-
+  
   // Now serialize each emission.  If we are loading, we must resize the vector
   // of emissions correctly.
-  if (Archive::is_loading::value)
-    emission.resize(transition.n_rows);
-
+  emission.resize(transition.n_rows);
   // Load the emissions; generate the correct name for each one.
-    ar & BOOST_SERIALIZATION_NVP(emission);
+  ar & BOOST_SERIALIZATION_NVP(emission);
+  
+  initialProxy = initial;
+  transitionProxy = transition;
+  logTransition = log(transition);
+  logInitial = log(initial);
+ }
+
+//! Serialize the HMM.
+template<typename Distribution>
+template<typename Archive>
+void HMM<Distribution>::save(Archive& ar, const unsigned int /* version */) const
+{
+  arma::mat transition = exp(logTransition);
+  arma::vec initial = exp(logInitial);
+  ar & BOOST_SERIALIZATION_NVP(dimensionality);
+  ar & BOOST_SERIALIZATION_NVP(tolerance);
+  ar & BOOST_SERIALIZATION_NVP(transition);
+  ar & BOOST_SERIALIZATION_NVP(initial);
+  ar & BOOST_SERIALIZATION_NVP(emission);
```

Comment:
Do you think it's possible to leave these `load()` and `save()` functions combined as `serialize()`?  I think it can be done to save some space, but maybe it's just a matter of taste:

```
  arma::mat transition, initial;

  if (Archive::is_saving::value)
  {
    transition = exp(logTransition);
    initial = exp(logInitial);
  }

  ar & BOOST_SERIALIZATION_NVP(dimensionality);
  ar & BOOST_SERIALIZATION_NVP(tolerance);
  ar & BOOST_SERIALIZATION_NVP(transition);
  ar & BOOST_SERIALIZATION_NVP(initial);

  if (Archive::is_loading::value)
  {
    emission.resize(transition.n_rows);
    initialProxy = std::move(initial);
    transitionProxy = std::move(transition);
    logInitial = log(initialProxy);
    logTransition = log(transitionProxy);
  }

  ar & BOOST_SERIALIZATION_NVP(emission);
```

(The `std::move()`s there also allow us to avoid a copy.)

Reply:
Yes, They can be merged. I think having two separate functions for `load` and `save` makes the code more readable and less error prone. It's up to you.

Reply:
I wanted to introduce the proxy variables with the minimum intrusion into the code so it can be removed later with minimum effort since they are just a workaround, hence I avoided using your suggestion for `std::move()` because it requires editing the other part of the code once you want to get rid of the proxies. With the way that I wrote the code, you just need to remove the proxies and not being worry about the rest. I guess for me easy maintenance and readability has higher priority but at the end it's up to you to make the decision.

Reply:
That's a fair point.  The places where the copies happen are definitely places that won't be called often, so although it's unnecessary, it won't be any kind of serious bottleneck.  So I'm okay leaving it as-is. :+1:  Two separate functions for `load` and `save` is fine too.